ibis/backends/base/regex.py

```diff
@@ -0,0 +1,98 @@
+import re
+
+
+def normalize(r):
+    """Normalize a regular expression by ensuring that it is wrapped with:
+    '^' and '$'
+
+    Parameters
+    ----------
+    r : str or Pattern
+        The pattern to normalize.
+
+    Returns
+    -------
+    p : Pattern
+        The compiled regex.
+    """
+    r = getattr(r, "pattern", r)
+    return re.compile('^' + r.lstrip('^').rstrip('$') + '$')
+
+
+class RegexDispatcher:
+    """
+    Regular Expression Dispatcher
+
+    >>> f = RegexDispatcher('f')
+
+    >>> f.register('\\d*')
+    ... def parse_int(s):
+    ...     return int(s)
+
+    >>> f.register('\\d*\\.\\d*')
+    ... def parse_float(s):
+    ...     return float(s)
+
+    Set priorities to break ties between multiple matches.
+    Default priority is set to 10
+
+    >>> f.register('\\w*', priority=9)
+    ... def parse_str(s):
+    ...     return s
+
+    >>> type(f('123'))
+    int
+
+    >>> type(f('123.456'))
+    float
+    """
+
+    def __init__(self, name):
+        self.name = name
+        self.funcs = {}
+        self.priorities = {}
+
+    def add(self, regex, func, priority=10):
+        self.funcs[normalize(regex)] = func
+        self.priorities[func] = priority
+
+    def register(self, regex, priority=10):
+        """Register a new handler in this regex dispatcher.
+
+        Parameters
+        ----------
+        regex : str or Pattern
```

Comment:
In general I think we should avoid the colon typing syntax in docstrings and just add the equivalent type annotations. `mkdocstrings` does a better job with type annotations than with the numpydoc type strings.